Sources/Fluent/Database/Database.swift

```diff
@@ -1,21 +1,23 @@
-import Foundation
-
 /**
     References a database with a single `Driver`.
     Statically maps `Model`s to `Database`s.
 */
-public class Database {
-    /**
-        The `Driver` powering this database. 
-        Responsible for executing queries.
-    */
+public class Database: Executor {
+    /// The `Driver` powering this database.
+    /// Responsible for executing queries.
     public let driver: Driver
+    
+    /// Maintains a pool of connections
+    /// one for each thread
+    public let threadConnectionPool: ThreadConnectionPool
 
-    /** 
-        Creates a `Database` with the supplied
-        `Driver`. This cannot be changed later.
-    */
+    /// Creates a `Database` with the supplied
+    /// `Driver`. This cannot be changed later.
     public init(_ driver: Driver) {
+        threadConnectionPool = ThreadConnectionPool(
+            makeConnection: driver.makeConnection,
+            maxConnections: 5
```

Comment:
In MongoKitten we have a pool that you reserve a connection at. If there is a connection available we use that (with some more advanced logic to select the best connection), otherwise we create a new connection and put it in the pool.

We have a reserve->return mechanism where you reserve a connection for an unspecified amount of queries/duration and you return it when you're done.

So almost always whenever you're executing a query you'd write something like this:

```swift
let connection = try reserveConnection()
defer { returnConnection(connection) }
// use connection for a query
```

We keep the connection reserved in one case, which are cursors. Because cursors can fetch data for one query more memory/response-time efficient by fetching the response in parts.
So if you have 5000 documents it'd fetch them in chunks rather than a whole.

Our pool expands to 100, which sounds crazy, but I think it's not. 100 connections means you can serve ~100 clients simultaneously. You could go down to 50, but I don't think I'd go lower.

In my case I've got first-hand experience with 50 and 100 connections as a difference. If you're running a website that gets a lot of API requests from it's clients you can quickly reach 50. Especially if you have an iPhone and/or web-app that do 3-5 API requests per view this goes down drastically sometimes for some medium/bigger sized users.